### ISSUE TITLE
Support replacing an OSD based on a metadataDevice that is shared with multiple OSDs on a node

### ISSUE THREAD

**Issue by travisn** [rook/rook#13240]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Rook should allow replacing a single OSD on a node where multiple OSDs are configured to share a metadataDevice. Currently all OSDs that share a metadataDevice would need to be replaced at the same time, which is much more disruptive and risky for fault tolerance and impact of backfill after the replacement of all the OSDs on the node at the same time. If we can support the single replacement of an OSD, it effectively improves the fault tolerance of replacing the OSD from a node-level to just a single OSD failure.

This issue does not affect [clusters on PVCs](https://rook.io/docs/rook/latest/CRDs/Cluster/pvc-cluster/#dedicated-metadata-and-wal-device-for-osd-on-pvc) since each metadataDevice is already only associated with a single data device.

**Background:**
A common configuration for increasing performance of OSDs is where a high performance metadataDevice (nvme/ssd) is configured as the db/wal, while an hdd is used for the data device. This allows lowering the cost of the infrastructure when nvme/ssds are not available for all devices. 

**What is use case behind this feature:**
Lower the cost and risk of replacing an OSD. Instead of replacing all OSDs on a node at the same time, a single OSD can be replaced. 

**Design Considerations**
When Rook provisions OSDs (in the OSD prepare job) for this scenario, Rook effectively runs a `c-v` command such as the following. In this example, there are three data devices with one metadataDevice.
```
ceph-volume lvm batch --prepare <deviceA> <deviceB> <deviceC> --db-devices <metadataDevice>
```

If the metadataDevice dies, all of the OSDs need to be replaced, there is no way around that. 

If one of the data devices dies, Rook should support just replacing that single device instead of wiping and replacing all OSDs at the same time. The first step is to [purge the OSD](https://rook.io/docs/rook/latest/Storage-Configuration/Advanced/ceph-osd-mgmt/#remove-an-osd), just as any other OSD would be purged. The second steps is then to create the new OSD in its place.

After discussing with @guits, `ceph-volume` does allow replacing a single OSD in this configuration, Rook only needs to design how to support it. The requirements include:
1. All data devices must be passed to `ceph-volume lvm batch`, even if only one device is to be replaced. `c-v` will detect that OSDs are still valid on the other devices, then go ahead and provision the one new OSD on the new data device. The section of the metadataDevice that was previously occupied by the purged OSD will be used for the new OSD.
2. There was a design change to `c-v` some time ago for batch mode that changed this scenario from using a single VG to a separate VG per OSD. If OSDs were created with the single VG, there will be additional manual steps required (using `c-v lvm migrate`) to accommodate this transition. 

It would be helpful to understand if we need to design for this issue, or if Rook users are not widely affected. 
@dimm0 @zhucan @microyahoo For your clusters do you see a single VG, or individual VGs for your OSDs on the metadataDevice? 

For the first design point, I'm thinking that if the OSD prepare job detects an available disk without an OSD when the metadataDevice is specified, Rook can call the `c-v` command above with all the devices in the list, even though they already have devices provisioned. This is different from Rook's pattern of skipping the OSD prepare if there is already an OSD existing on a device. I haven't looked through the code yet, but I believe this issue can be reasonably scoped so it will be a bug fix and not a big design change.

**Comment by zhucan** [comment]
@travisn under nvme cluster, the metadata store and data store use the same device; but under hdd cluster, we want to division one or more nvme device as the metadata device of osd.

**Comment by dimm0** [comment]
I see a single VG on the NVME

**Comment by microyahoo** [comment]
Our scenario is the same as what @zhucan mentioned above. And It would be best if rook could implement the following two functions:
1. If one of the data devices dies, support just replacing that single device instead of wiping and replacing all OSDs at the same time as you mentioned above @travisn 
2. support to specify disk partition or lv as metadata device, osd and partition or lv have a 1:1 relationship, which are not shared with multiple osds on one node. In other words, we can divide multiple disks into multiple partitions on one node, and each OSD uses one of the partitions.

**Comment by travisn** [comment]
> I see a single VG on the NVME

@guits What are the extra steps in this case?

**Comment by travisn** [comment]
@microyahoo @zhucan Are you saying you want to use multiple metadataDevices? For example, if you have two nvme devices and six hdds, you want three hdds for each nmve metadata device? In any case, please open a new issue for that request, so we can keep this issue focused on osd replacement.

**Comment by zhucan** [comment]
@travisn yes, we want to use multi metadataDevices，For exampe , we have two nvme devices and 6 hdds, we need to divsion every nvme devices to three partition as the metadata devices of different hdd devices. and create osd.

**Comment by microyahoo** [comment]
> @microyahoo @zhucan Are you saying you want to use multiple metadataDevices? For example, if you have two nvme devices and size hdds, you want three hdds for each nmve metadata device? In any case, please open a new issue for that request, so we can keep this issue focused on osd replacement.

ok, we can first solve osd replacement, then to support multiple metadata device in another thread.

**Comment by travisn** [comment]
> @travisn yes, we want to use multi metadataDevices，For exampe , we have two nvme devices and 6 hdds, we need to divsion every nvme devices to three partition as the metadata devices of different hdd devices. and create osd.

@zhucan Please create a new issue for that

**Comment by zhucan** [comment]
@travisn done

**Comment by microyahoo** [comment]
As [document](https://rook.io/docs/rook/v1.12/CRDs/Cluster/host-cluster/#specific-nodes-and-devices) mentioned, configuration can be specified at cluster level, the node level or device level config like below. 

cluster level:
```yaml
storage:
    useAllNodes: false
    useAllDevices: false
    deviceFilter:
    config:
      metadataDevice: /dev/nmve01
    nodes:
    - name: "172.17.4.201"
      devices:             
      - name: "sdb"
      - name: "sdc1"
```
node level:
```yaml
storage:
    useAllNodes: false
    useAllDevices: false
    deviceFilter:
    nodes:
    - name: "172.17.4.201"
      config:
        metadataDevice: /dev/nmve01
      devices:             
      - name: "sdb"
      - name: "sdc1"
```
device level:
```yaml
storage:
    useAllNodes: false
    useAllDevices: false
    deviceFilter:
    nodes:
    - name: "172.17.4.201"
      devices:             
      - name: "sda"
        config:
          metadataDevice: "/dev/sdc1"
      - name: "sdb"
        config:
          metadataDevice: "/dev/sdc2"
```
We can easily replace osd based on the metadata device at the device level, but how to deal with the cluster level and node level? @travisn @satoru-takeuchi

**Comment by satoru-takeuchi** [comment]
@microyahoo 

> We can easily replace osd based on the metadata device at the device level, but how to deal with the cluster level and node level?

This is what exactly this feature request will do. Correct me if I misunderstood something.

**Comment by microyahoo** [comment]
> @microyahoo
> 
> > We can easily replace osd based on the metadata device at the device level, but how to deal with the cluster level and node level?
> 
> This is what exactly this feature request will do. Correct me if I misunderstood something.

yes, it's hard to replace osd in cluster level and node level, because the CRD doesn't record the corresponding metadata device/partition of data device.

**Comment by sp98** [comment]
@travisn  Started looking into this now.  Correct me if my understanding is wrong. 

- User prepares OSDs using `ceph-volume lvm batch --prepare <deviceA> <deviceB> <deviceC> --db-devices <metadataDevice>` 
- OSD on `deviceA` needs to be replaced due to some reason, say, backing disk gone bad. 

Following steps can be used to replace the OSD without affecting the other OSDs using same `metadataDeivice` 
1.  `purge` the OSD to be removed. 
2. Clean the backing disk `deviceA`
3.  Remove the db `lv` created for `deviceA` 
4. Once the OSD is ready to be redeployed, run the same OSD prepare command - `ceph-volume lvm batch --prepare <deviceA> <deviceB> <deviceC> --db-devices <metadataDevice>` 

Steps 1, 2, 3 requires manual intervention by the user.  The rook operator needs to be down during step `2` and `3`.  Once OSD is ready to be redeployed, the operator should be restarted, the prepare pod should identify that OSD for `deviceA` is not configured and it would run the same command `ceph-volume lvm batch --prepare <deviceA> <deviceB> <deviceC> --db-devices <metadataDevice>` to prepare the OSD again. 

Is that the correct expectation?

**Comment by travisn** [comment]
> Steps 1, 2, 3 requires manual intervention by the user. The rook operator needs to be down during step `2` and `3`. Once OSD is ready to be redeployed, the operator would be restarted, the prepare pod would identify that OSD for `deviceA` is not configured and it would run the same command `ceph-volume lvm batch --prepare <deviceA> <deviceB> <deviceC> --db-devices <metadataDevice>` to prepare the OSD again.

Ok great to hear that ceph-volume supports provisioning just a single OSD in this case. 

I believe the OSD prepare job currently skips the provisioning of OSDs when using the metadata device if any of them have been previously created. Instead of skipping, now we could just run the command every time the prepare pod runs. Then we just need to confirm if no OSDs have been purged then it would be a no-op, and if one more more OSDs have been purged then the missing ones would be created.

**Comment by sstrk** [comment]
This feature is something which we would love to see realized. Please do not autoclose this issue.

**Comment by sp98** [comment]
> > Steps 1, 2, 3 requires manual intervention by the user. The rook operator needs to be down during step `2` and `3`. Once OSD is ready to be redeployed, the operator would be restarted, the prepare pod would identify that OSD for `deviceA` is not configured and it would run the same command `ceph-volume lvm batch --prepare <deviceA> <deviceB> <deviceC> --db-devices <metadataDevice>` to prepare the OSD again.
> 
> Ok great to hear that ceph-volume supports provisioning just a single OSD in this case.
> 
> I believe the OSD prepare job currently skips the provisioning of OSDs when using the metadata device if any of them have been previously created. Instead of skipping, now we could just run the command every time the prepare pod runs. Then we just need to confirm if no OSDs have been purged then it would be a no-op, and if one more more OSDs have been purged then the missing ones would be created.

prepare pod runs some filters on the discovered device, for example, if they have filesystem.  So it may not be very easy to run the same command to prepare the device. 
Checking if we can somehow get the devices  already configured using the `c-v` command.

**Comment by sp98** [comment]
One of the possible option is to use a configmap to store the ceph-volume configuration of the devices. But that won't work for the existing clusters, IMO.

**Comment by travisn** [comment]
Hopefully we don't need additional metadata for this scenario. In my osd-prepare log in minikube at the end it prints out the existing OSDs that were found such as:
```
2024-05-10 21:51:39.865724 I | cephosd: 3 ceph-volume raw osd devices configured on this node
2024-05-10 21:51:39.865793 I | cephosd: devices = [{ID:1 Cluster:ceph UUID:677dbe0c-afc6-447f-8b1d-439bdd6544d3 DevicePartUUID: DeviceClass:hdd 
BlockPath:/dev/vdb MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=minikube LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity: Encrypted:false ExportService:false NodeName: PVCName:} {ID:0 Cluster:ceph UUID:aebc9bb0-0181-4a5b-a437-7bb6732301ab DevicePartUUID: DeviceClass:hdd BlockPath:/dev/vdd MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=minikube LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity: Encrypted:false ExportService:false NodeName: PVCName:} {ID:2 Cluster:ceph UUID:0b9323cd-8874-48f0-8abe-ac27bed34894 DevicePartUUID: DeviceClass:hdd BlockPath:/dev/vdc MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=minikube LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity: Encrypted:false ExportService:false NodeName: PVCName:}]
```

In particular, notice that the `BlockPath:/dev/vdb`. Perhaps we can use this property from the existing OSDs.

**Comment by microyahoo** [comment]
> Hopefully we don't need additional metadata for this scenario. In my osd-prepare log in minikube at the end it prints out the existing OSDs that were found such as:
> 
> ```
> 2024-05-10 21:51:39.865724 I | cephosd: 3 ceph-volume raw osd devices configured on this node
> 2024-05-10 21:51:39.865793 I | cephosd: devices = [{ID:1 Cluster:ceph UUID:677dbe0c-afc6-447f-8b1d-439bdd6544d3 DevicePartUUID: DeviceClass:hdd 
> BlockPath:/dev/vdb MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=minikube LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity: Encrypted:false ExportService:false NodeName: PVCName:} {ID:0 Cluster:ceph UUID:aebc9bb0-0181-4a5b-a437-7bb6732301ab DevicePartUUID: DeviceClass:hdd BlockPath:/dev/vdd MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=minikube LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity: Encrypted:false ExportService:false NodeName: PVCName:} {ID:2 Cluster:ceph UUID:0b9323cd-8874-48f0-8abe-ac27bed34894 DevicePartUUID: DeviceClass:hdd BlockPath:/dev/vdc MetadataPath: WalPath: SkipLVRelease:true Location:root=default host=minikube LVBackedPV:false CVMode:raw Store:bluestore TopologyAffinity: Encrypted:false ExportService:false NodeName: PVCName:}]
> ```
> 
> In particular, notice that the `BlockPath:/dev/vdb`. Perhaps we can use this property from the existing OSDs.

yes, osd-prepare log records existing osd devices, but we will have no way to obtain this information if the user deletes the corresponding osd-prepare pods.

**Comment by travisn** [comment]
> yes, osd-prepare log records existing osd devices, but we will have no way to obtain this information if the user deletes the corresponding osd-prepare pods.

My assumption is that to support this we will have some code running inside the osd prepare job, so we can extract that info from ceph-volume. No need to read the osd prepare log by some external process.

**Comment by sstrk** [comment]
Please do not autoclose. Thanks!

**Comment by sstrk** [comment]
Please do not autoclose. Thanks!

**Comment by huapox** [comment]
+1

**Comment by sumitarora2786** [comment]
we have a similar setup, 12 HDD OSDs share one NVMe drive meant for metadata, there is a VG for each OSD, but single VG for all the metadata LV partitions, while we followed the procedure to replace only single HDD by using ceph-volume, there are few findings:

- for a HDD drive replacement, while we try to use the existing metadata LV to create the OSD on a new HDD, we see some issues afterwords.

ceph-volume lvm prepare --bluestore --dmcrypt --data /dev/sdk --block.db /dev/xxxxxxxxxxxxx --crush-device-class hdd

Running ceph config-key get dm-crypt/osd/xxxx/luks
Running command: /usr/bin/ceph --cluster ceph --name client.osd-lockbox.xxx--keyring /var/lib/ceph/osd/ceph-80/lockbox.keyring config-key get dm-crypt/osd/xxx/luks
Running command: /usr/sbin/cryptsetup --key-size 512 --key-file - --allow-discards luksOpen /dev/xxx
 stderr: Device xxx already exists.
Running command: /usr/sbin/cryptsetup --key-size 512 --key-file - --allow-discards luksOpen /dev/xxx
 stderr: No key available with this passphrase.


to correct, we removed the LV partition for given OSD from the VG and created a LV partition again, then used it to create the OSD via ceph-volume.

**Comment by virtualluke** [comment]
Any update on this? This would be very helpful.

**Comment by arttor** [comment]
@virtualluke unfortunately i have not started any work on it and will be able to start only in a couple of weeks. If there are someone who has time to work on it now, we can reassign it.

**Comment by veshant** [comment]
I think this would also be useful for adding devices to a node using an existing metadata device.

**Comment by veshant** [comment]
just wondering, is anyone working on this?

**Comment by laralar** [comment]
@sumitarora2786 
We have a similar issue as you, we have added 2x SATA SSD on the server and 10 HDD, 5 x HDD per SSD

```
    nodes:
      - name: "dmz01"
        devices:
          - name: "/dev/disk/by-id/scsi-SATA_WDC_WUS721010AL_VH1UJJ5M"
            config:
              metadataDevice: "/dev/disk/by-id/scsi-SATA_SAMSUNG_MZ7LH960_S45NNA1T819594"   # SSD1
...
          - name: "/dev/disk/by-id/scsi-SATA_WDC_WD80PUZX-64N_VKGX9B9X"
            config:
              metadataDevice: "/dev/disk/by-id/scsi-SATA_SSDSC2KB960GZR_PHYI4050030A960CGN"   # SSD2
```

A HDD has failed and we have: 
1. removed the LV on the SSD
2. manually created a new LV
3. manually created the osd with ceph-volume, it shows UP in the ceph datshboard, but rook is not recognizing it and not creating the pod.

Any hint on how make rook recognize the OSD  and create the pod?

thanks,
Luis

**Comment by sumitarora2786** [comment]
can you access: https://github.com/cobaltcore-dev/cloud-storage/blob/main/docs/operations/replace-single-osd-example.md

**Comment by laralar** [comment]
I get a 404

**Comment by laralar** [comment]
I was able to fix the issue and replace the OSD manually on a shared SSD with multiple LV for different OSDs.

1. scale down the operator

2. purge the failed OSD

3. delete any remaining directories in the host linked to the DB SSD/NVME, delete the auth key in ceph for the osd if needed.

4. delete the DB LV in the SSD/NVME and create it again
```
lvremove ......
VGNAME=ceph-XXXXX # already exists, find it out.
LVNAME=osd-block-$(uuidgen)
lvcreate -l 100%FREE -n $LVNAME $VGNAME # in my case I use all the remaining space, adjust accordingly
```

5. create the PV, VG, LV on the HDD # I needed to do this because I am using rook inside suse harvester, meaning I don't have access to ceph commands in the host, I have to create a docker container for running the ceph commands, I believe if can run the ceph-volume lvm prepare command in your host directly this is not needed.  as passing the RAW device to the prepare will create the LV.  In my case, it was getting stuck in the docker container, I had to create the LV for the data block outside.
```
pvcreate /dev/sdr
VGNAME=ceph-$(uuidgen)
LVNAME=osd-block-$(uuidgen)
vgcreate $VGNAME /dev/sdr
lvcreate -l 100%FREE -n $LVNAME $VGNAME
```

6. create a docker container for being able to run the ceph commands, even though I am using 19.2.0, I had to create an image with 19.2.2, because the ceph-volume command was giving me an error trying to prepare the OSD, not needed if you can install/run ceph-volume in your host.
```
docker run -it --rm --privileged   -v /dev:/dev   -v /run/udev:/run/udev   -v /var/lib/rook/rook-ceph:/var/lib/rook/rook-ceph   -v /var/lib/ceph:/var/lib/ceph   -v /var/lib/rook/rook-ceph/rook-ceph.config:/etc/ceph/ceph.conf:ro   -v /var/lib/rook/rook-ceph/client.admin.keyring:/etc/ceph/ceph.client.admin.keyring:ro   quay.io/ceph/ceph:v19.2.2  bash
``` 

7. setup the bootstrap auth keys inside the container, not needed if you can install/run ceph-volume in your host.
```
set -e
mkdir -p /var/lib/ceph/bootstrap-osd
key=$(ceph -c /etc/ceph/ceph.conf \
           --name client.admin \
           --keyring /etc/ceph/ceph.client.admin.keyring \
           auth get-key client.bootstrap-osd)
printf "[client.bootstrap-osd]\n\tkey = %s\n" "$key" > /var/lib/ceph/bootstrap-osd/ceph.keyring
chmod 0400 /var/lib/ceph/bootstrap-osd/ceph.keyring
echo "✅ bootstrap keyring created"
```

8. prepare the OSD, in my case inside the container.
```
ceph-volume lvm prepare \
  --bluestore \
  --data /dev/ceph-518cfa35-956f-4dc4-ac4e-1f6056d7f2d7/osd-block-9304c5e1-f7c6-46b1-bebb-638c63d5754f \
  --block.db /dev/ceph-a6791ca0-d67a-4d12-bcc3-db18b44f1fff/osd-db-b9e4b57a-0b0e-4be1-a283-92e639fb9946
echo "✅ OSD prepared"
```

9. scale up rook. the OSD pod should come up.

**Comment by virtualluke** [comment]
Is there any update for this?

**Comment by davidturner-sie** [comment]
> After discussing with @guits, ceph-volume does allow replacing a single OSD in this configuration, Rook only needs to design how to support it. The requirements include:
>
> All data devices must be passed to ceph-volume lvm batch, even if only one device is to be replaced. c-v will detect that OSDs are still valid on the other devices, then go ahead and provision the one new OSD on the new data device. The section of the metadataDevice that was previously occupied by the purged OSD will be used for the new OSD.

@travisn If it's just a matter of tracking which devices are using a shared SSD, the command to find the list is really simple. Given OSDs on `osd3.mycluster.lan` using `nvme0n1` as their shared device, this command gives you a list of every block device using it.
`ceph osd metadata | jq -r '.[] | select(.hostname =="osd3.mycluster.lan") | select(.bluefs_db_devices == "nvme0n1") | .bluestore_bdev_devices'`

If this isn't the main hurdle, what is?

**Comment by travisn** [comment]
> If this isn't the main hurdle, what is?

The osd provisioning code is complex. I spent a bit of time today, still trying to track it down...

**Comment by virtualluke** [comment]
Any update?  This feature would be really nice to have.

**Comment by mlazowik** [comment]
Not stale

**Comment by jasonburris-sie** [comment]
Please keep this one open.

**Comment by travisn** [comment]
Mistakenly closed, the implementation is still coming with #17953

**Comment by travisn** [comment]
This feature will be included in Rook v1.21...
